dialogue-core/src/main/java/com/palantir/dialogue/core/ClientPoolImpl.java

```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    @Override
+    public <T> T get(Class<T> dialogueInterface, Listenable<DialogueConfig> config) {
+        Channel channel = smartChannel(config);
+        // Note: if you call this many times, you'll get entirely independent 'smart channel' instances, which means
+        // they each have their own idea about blacklisting / concurrency limiting etc.
+        return conjure(dialogueInterface, channel, runtime);
+    }
+
+    @Override
+    public Channel smartChannel(Listenable<DialogueConfig> config) {
+        // This is a naive approach to live reloading, as it throws away all kinds of useful state (e.g. active request
+        // count, blacklisting info etc).
+        return RefreshingChannelFactory.RefreshingChannel.create(config::getListenableCurrentValue, conf -> {
+            List<Channel> channels = conf.uris().stream()
+                    .map(uri -> {
+                        // important that this re-uses resources under the hood, as it gets called often!
+                        return rawHttpChannel(uri, config);
+                    })
+                    .collect(Collectors.toList());
+
+            return Channels.create(channels, conf.userAgent, conf.legacyClientConfiguration);
+        });
+    }
+
+    @Override
+    public Channel rawHttpChannel(String uri, Listenable<DialogueConfig> config) {
+        Class<? extends HttpChannelFactory> factoryClazz = config.getListenableCurrentValue().httpChannelFactory;
+
+        config.subscribe(() -> {
+            if (config.getListenableCurrentValue().httpChannelFactory != factoryClazz) {
+                log.warn("Live-reloading the *type* of underlying http channel is not currently supported");
+            }
+        });
+
+        HttpChannelFactory channelFactory = getOnlyEnumConstant(factoryClazz);
+
+        // by passing in sharedResources, we're enable to avoid re-creating the expensive underlying clients
+        return channelFactory.construct(uri, config, sharedResources);
+    }
+
+    /** Just give me a working implementation of the provided conjure-generated interface. */
+    @VisibleForTesting
+    static <T> T conjure(Class<T> dialogueInterface, Channel smartChannel, ConjureRuntime runtime) {
+        ConstructUsing annotation = dialogueInterface.getDeclaredAnnotation(ConstructUsing.class);
+        Preconditions.checkNotNull(
+                annotation,
+                "@ConstructUsing annotation must be present on interface",
+                SafeArg.of("interface", dialogueInterface.getName()));
+
+        Class<?> factoryClass = annotation.value();
+        // this is safe because of the ConstructUsing annotation's type parameters
+        Factory<T> factory = (Factory<T>) getOnlyEnumConstant(factoryClass);
+        return factory.construct(smartChannel, runtime);
+    }
+
+    private static <F> F getOnlyEnumConstant(Class<F> factoryClass) {
```

Comment:
```suggestion
    private static <F extends Enum<F>> F getOnlyEnumConstant(Class<F> factoryClass) {
```

Reply:
The only way I can get this to work is if I then make the argument `Class<?>`, which seems a bit less nice?

![image](https://user-images.githubusercontent.com/3473798/75039646-00590500-54b1-11ea-8888-c91f8ab5520c.png)